src/services/refactors/extractMethod.ts

```diff
@@ -0,0 +1,1154 @@
+                range.facts & RangeFacts.IsGenerator ? createToken(SyntaxKind.AsteriskToken) : undefined,
+                functionName,
+                /*typeParameters*/[],
+                parameters,
+                returnType,
+                body
+            );
+        }
+
+        const changeTracker = textChanges.ChangeTracker.fromCodeFixContext(context);
+        // insert function at the end of the scope
+        changeTracker.insertNodeBefore(context.file, scope.getLastToken(), newFunction, { prefix: context.newLineCharacter, suffix: context.newLineCharacter });
+
+        const newNodes: Node[] = [];
+        // replace range with function call
+        let call: Expression = createCall(
+            isClassLike(scope) ? createPropertyAccess(range.facts & RangeFacts.InStaticRegion ? createIdentifier(scope.name.getText()) : createThis(), functionReference) : functionReference,
+            /*typeArguments*/ undefined,
+            callArguments);
+        if (range.facts & RangeFacts.IsGenerator) {
+            call = createYield(createToken(SyntaxKind.AsteriskToken), call);
+        }
+        if (range.facts & RangeFacts.IsAsyncFunction) {
+            call = createAwait(call);
+        }
+
+        if (writes) {
+            if (returnValueProperty) {
+                // has both writes and return, need to create variable declaration to hold return value;
+                newNodes.push(createVariableStatement(
+                    /*modifiers*/ undefined,
+                    [createVariableDeclaration(returnValueProperty, createKeywordTypeNode(SyntaxKind.AnyKeyword))]
+                ));
+            }
+
+            const assignments = getPropertyAssignmentsForWrites(writes);
+            if (returnValueProperty) {
+                assignments.unshift(createShorthandPropertyAssignment(returnValueProperty));
+            }
+
+            // propagate writes back
+            if (assignments.length === 1) {
+                if (returnValueProperty) {
+                    newNodes.push(createReturn(createIdentifier(returnValueProperty)));
+                }
+                else {
+                    newNodes.push(createStatement(createBinary(assignments[0].name, SyntaxKind.EqualsToken, call)));
+                }
+            }
+            else {
+                // emit e.g.
+                //   { a, b, __return } = newFunction(a, b);
+                //   return __return;
+                newNodes.push(createStatement(createBinary(createObjectLiteral(assignments), SyntaxKind.EqualsToken, call)));
+                if (returnValueProperty) {
+                    newNodes.push(createReturn(createIdentifier(returnValueProperty)));
+                }
+            }
+        }
+        else {
+            if (range.facts & RangeFacts.HasReturn) {
+                newNodes.push(createReturn(call));
+            }
+            else if (isArray(range.range)) {

```

Comment:
>isArray(range.range) [](start = 21, length = 20)

Consider extracting a local for this.  It has the opposite sense, but "isExpression" might be the most readable.